docs/design/paginators.md

```diff
@@ -0,0 +1,297 @@
+
+fun<T> SdkAsyncIterable<Iterable<T>>.flatten(): SdkAsyncIterable<T> = TODO("not-implemented")
+fun <T> SdkAsyncIterable<T?>.filterNotNull(): SdkAsyncIterable<T> = TODO("not-implemented")
+
+suspend fun<T> SdkAsyncIterable<T>.toList(): List<T> = TODO("not-implemented")
+suspend fun<T> SdkAsyncIterable<T>.toSet(): Set<T> = TODO("not-implemented")
+
+/**
+ * Terminal operator that collects the given iterable and calls [action] for each result
+ */
+inline fun <T> SdkAsyncIterable<T>.collect(crossinline action: suspend (T) -> Unit):Unit = TODO()
+```
+
+Pagination by default will always generate an `SdkAsyncIterable` over the normal operation output type. 
+If `items` is specified then the generated paginator will be “specialized” to have an additional field that iterates 
+only over the member targeted by `items`. The targeted type will be flattened for the user automatically.
+
+An example of this is demonstrated below:
+
+
+```kotlin
+class ListFunctionsPaginator(
+    private val client: LambdaClient,
+    private val initialRequest: ListFunctionsRequest
+): SdkAsyncIterable<ListFunctionsResponse>, SdkAsyncIterator<ListFunctionsResponse> {
+   private var cursor: String? = null
+   private var isFirstPage: Boolean = true
+
+   override operator fun iterator(): SdkAsyncIterator<ListFunctionsResponse> = this
+
+   override operator fun hasNext(): Boolean
+           = isFirstPage || (cursor?.isNotEmpty() ?: false)
+
+   override suspend operator fun next(): ListFunctionsResponse {
+      if (!hasNext()) throw NoSuchElementException("no pages remaining")
+
+      val req = initialRequest.copy {
+         this.marker = cursor
+      }
+
+      val result = client.listFunctions(req)
+      isFirstPage = false
+      cursor = result.nextMarker
+      return result
+   }
+
+   // "functions" is the member targeted by the trait's `items` property
+   val functions: SdkAsyncIterable<FunctionConfiguration> = mapNotNull(ListFunctionsResponse::functions).flatten()
+}
+```
+
+The shape targeted by `inputToken` is mapped to the `cursor` field. Each iteration the `cursor` is updated with the output field targeted by the `outputToken`.
+
+
+### Creating a Paginator
+
+Each operation that supports pagination would receive an extension function to create a paginator. Using the model in the introduction would produce the following:
+
+
+```kotlin
+// file: Paginators.kt
+package aws.sdk.kotlin.services.lambda
+
+import aws.sdk.kotlin.services.lambda.model.FunctionConfiguration
+import aws.sdk.kotlin.services.lambda.model.ListFunctionsRequest
+
+fun LambdaClient.listFunctionsPaginated(request: ListFunctionsRequest): ListFunctionsPaginator
+    = ListFunctionsPaginator(this, request)
+```
+
+
+
+### Example Usage - Manual Pagination
+
+An example of driving a paginator manually and processing results:
+
+```kotlin
+suspend fun rawPaginationExample(client: LambdaClient) {
+    val req = ListFunctionsRequest{}
```

Comment:
nit

whitespace

Reply:
I'm not sure what you're suggesting, the whitespace looks good to my 👀